File: walletcore/src/main/java/com/mycelium/wapi/wallet/ExportableAccount.java
package com.mycelium.wapi.wallet;


import com.google.common.base.Optional;
import com.mrd.bitlib.crypto.BipDerivationType;

import java.io.Serializable;
import java.util.Map;

public interface ExportableAccount {
   class Data implements Serializable {
      public final Optional<String> privateData;
      private Map<BipDerivationType, String> privateDataMap;
      public final Optional<String> publicData;
      private Map<BipDerivationType, String> publicDataMap;

      public Data(Optional<String> privateData, Optional<String> publicData) {
         this.privateData = privateData;
         this.publicData = publicData;
      }

      public Data(Map<BipDerivationType, String> privateDataMap, Map<BipDerivationType, String> publicDataMap) {
         this.privateData = Optional.of(privateDataMap.get(privateDataMap.keySet().iterator().next()));
         this.privateDataMap  = privateDataMap;
         this.publicData = Optional.of(publicDataMap.get(publicDataMap.keySet().iterator().next()));
         this.publicDataMap = publicDataMap;
      }

      public Map<BipDerivationType, String> getPrivateDataMap() {
         return privateDataMap;
      }

      public Map<BipDerivationType, String> getPublicDataMap() {
         return publicDataMap;
      }
   }

   Data getExportData(KeyCipher cipher);
}

File: walletcore/src/main/java/com/mycelium/wapi/wallet/ExportableAccount.kt
package com.mycelium.wapi.wallet


import com.google.common.base.Optional
import com.mrd.bitlib.crypto.BipDerivationType

import java.io.Serializable

interface ExportableAccount {
    class Data : Serializable {
        val privateData: Optional<String>
        var privateDataMap: Map<BipDerivationType, String>? = null
        var publicDataMap: Map<BipDerivationType, String>? = null

        constructor(privateData: Optional<String>, publicDataMap: Map<BipDerivationType, String>?) {
            this.privateData = privateData
            this.publicDataMap = publicDataMap
        }

        constructor(privateDataMap: Map<BipDerivationType, String>?, publicDataMap: Map<BipDerivationType, String>?) {
            this.privateData = Optional.fromNullable(privateDataMap?.get(privateDataMap.keys.iterator().next()))
            this.privateDataMap = privateDataMap
            this.publicDataMap = publicDataMap
        }
    }

    fun getExportData(cipher: KeyCipher): Data
}
